src/wallet/rpcdisclosure.cpp

```diff
@@ -0,0 +1,296 @@
+        );
+
+    if (!fEnablePaymentDisclosure) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error: payment disclosure is disabled.");
+    }
+
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+
+    EnsureWalletIsUnlocked();
+
+    // Check wallet knows about txid
+    string txid = params[0].get_str();
+    uint256 hash;
+    hash.SetHex(txid);
+
+    CTransaction tx;
+    uint256 hashBlock;
+
+    // Check txid has been seen
+    if (!GetTransaction(hash, tx, hashBlock, true)) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "No information available about transaction");
+    }
+
+    // Check tx has been confirmed
+    if (hashBlock.IsNull()) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Transaction has not been confirmed yet");
+    }
+
+    // Check is mine
+    if (!pwalletMain->mapWallet.count(hash)) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Transaction does not belong to the wallet");
+    }
+    const CWalletTx& wtx = pwalletMain->mapWallet[hash];
+
+    // Check if shielded tx
+    if (wtx.vjoinsplit.empty()) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Transaction is not a shielded transaction");        
+    }
+
+    // Check js_index
+    int js_index = params[1].get_int();
+    if (js_index < 0 || js_index >= wtx.vjoinsplit.size()) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid js_index");
+    }
+
+    // Check output_index
+    int output_index = params[2].get_int();
+    if (output_index < 0 || output_index >= ZC_NUM_JS_OUTPUTS) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid output_index");
+    }
+
+    // Get message if it exists
+    string msg;
+    if (params.size() == 4) {
+        msg = params[3].get_str();
+    }
+
+    // Create PaymentDisclosureKey
+    PaymentDisclosureKey key = {hash, (size_t)js_index, (uint8_t)output_index };
+
+    // TODO: In future, perhaps init the DB in init.cpp
+    shared_ptr<PaymentDisclosureDB> db = PaymentDisclosureDB::sharedInstance();
+    PaymentDisclosureInfo info;
+    if (!db->Get(key, info)) {
+        throw JSONRPCError(RPC_DATABASE_ERROR, "Could not find payment disclosure info for the given joinsplit output");
+    }
+
+    PaymentDisclosure pd( wtx.joinSplitPubKey, key, info, msg );
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << pd;
+    string strHex = HexStr(ss.begin(), ss.end());
+    return strHex;
+}
+
+
+
+/**
+ * RPC call to validate a payment disclosure data blob.
+ */
+UniValue z_validatepaymentdisclosure(const UniValue& params, bool fHelp)
+{
+    if (!EnsureWalletIsAvailable(fHelp))
+        return NullUniValue;
+
+    auto fEnablePaymentDisclosure = fExperimentalMode && GetBoolArg("-paymentdisclosure", false);
+    string strPaymentDisclosureDisabledMsg = "";
+    if (!fEnablePaymentDisclosure) {
+        strPaymentDisclosureDisabledMsg = "\nWARNING: Payment disclosure is curretly DISABLED. This call always fails.\n";
```

Comment:
Have the message say how to enable it.

Reply:
We don't tell users how to enable other features such as `-developerencrypewallet` so I'm inclined to leave this as is for now.  I've filed a general ticket for this #2728